File: tools/validation/extract_diagrams.mjs
#!/usr/bin/env node
/**
 * Extract mermaid diagrams from Cypress test files
 *
 * Usage: node extract_diagrams.mjs /path/to/mermaid/cypress > cypress_diagrams.json
 */

import { readFileSync, readdirSync, statSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function findTestFiles(dir) {
  const files = [];

  function walk(currentDir) {
    const entries = readdirSync(currentDir);
    for (const entry of entries) {
      const fullPath = join(currentDir, entry);
      const stat = statSync(fullPath);
      if (stat.isDirectory()) {
        walk(fullPath);
      } else if (entry.endsWith('.spec.js') || entry.endsWith('.spec.ts')) {
        files.push(fullPath);
      }
    }
  }

  walk(dir);
  return files;
}

// Process JavaScript escape sequences in a string
function unescapeJsString(str) {
  // Process escape sequences as JavaScript would
  // Note: Order matters - process \\ before other escapes to avoid double processing
  return str
    .replace(/\\\\/g, '\x00')     // Temporarily replace \\ with a placeholder
    .replace(/\\n/g, '\n')
    .replace(/\\r/g, '\r')
    .replace(/\\t/g, '\t')
    .replace(/\\'/g, "'")
    .replace(/\\"/g, '"')
    .replace(/\\`/g, '`')          // Handle escaped backticks in template literals
    .replace(/\x00/g, '\\');       // Restore backslashes
}

function extractDiagrams(content, sourceFile) {
  const diagrams = [];

  // Match it(...) or describe(...) blocks with imgSnapshotTest
  // We need to find test names and their corresponding diagram strings

  // Pattern to match: it('test name', ... `diagram content` ...)
  // Note: We need to handle escaped backticks (\`) inside the template literal
  const itBlockRegex = /it\s*\(\s*['"]([^'"]+)['"]\s*,\s*(?:async\s*)?\(\s*\)\s*=>\s*\{[^}]*imgSnapshotTest\s*\(\s*`((?:[^`\\]|\\.)*)`/g;

  let match;
  while ((match = itBlockRegex.exec(content)) !== null) {
    const testName = match[1];
    let diagram = match[2];

    // The diagram in the JS file is a template literal read as raw text
    // We need to process escape sequences to get the actual content
    diagram = unescapeJsString(diagram);
    diagram = diagram.trim();

    // Detect diagram type from content
    let diagramType = 'unknown';
    if (diagram.includes('sequenceDiagram')) diagramType = 'sequence';
    else if (diagram.includes('classDiagram')) diagramType = 'class';
    else if (/^(graph|flowchart)/m.test(diagram)) diagramType = 'flowchart';
    else if (diagram.includes('stateDiagram')) diagramType = 'state';
    else if (diagram.includes('erDiagram')) diagramType = 'er';
    else if (diagram.includes('journey')) diagramType = 'journey';
    else if (diagram.includes('gantt')) diagramType = 'gantt';
    else if (diagram.includes('pie')) diagramType = 'pie';
    else if (diagram.includes('gitGraph')) diagramType = 'git';
    else if (diagram.includes('mindmap')) diagramType = 'mindmap';
    else if (diagram.includes('timeline')) diagramType = 'timeline';
    else if (diagram.includes('quadrantChart')) diagramType = 'quadrant';
    else if (diagram.includes('sankey')) diagramType = 'sankey';
    else if (diagram.includes('xychart')) diagramType = 'xychart';
    else if (diagram.includes('block-beta')) diagramType = 'block';
    else if (diagram.includes('packet-beta')) diagramType = 'packet';
    else if (diagram.includes('architecture')) diagramType = 'architecture';
    else if (diagram.includes('requirementDiagram')) diagramType = 'requirement';
    else if (diagram.includes('C4')) diagramType = 'c4';
    else if (diagram.includes('kanban')) diagramType = 'kanban';
    else if (diagram.includes('treemap')) diagramType = 'treemap';
    else if (diagram.includes('radar')) diagramType = 'radar';

    diagrams.push({
      test_name: testName,
      diagram_type: diagramType,
      diagram: diagram,
      source_file: sourceFile
    });
  }

  return diagrams;
}

async function main() {
  const mermaidCypressDir = process.argv[2] || '/Users/btucker/projects/mermaid/cypress';

  console.error(`Scanning ${mermaidCypressDir} for test files...`);

  const testFiles = findTestFiles(mermaidCypressDir);
  console.error(`Found ${testFiles.length} test files`);

  const allDiagrams = [];

  for (const file of testFiles) {
    try {
      const content = readFileSync(file, 'utf-8');
      const diagrams = extractDiagrams(content, file);
      allDiagrams.push(...diagrams);
    } catch (error) {
      console.error(`Error processing ${file}: ${error.message}`);
    }
  }

  console.error(`Extracted ${allDiagrams.length} diagrams`);

  // Output as JSON
  const output = {
    count: allDiagrams.length,
    diagrams: allDiagrams
  };

  console.log(JSON.stringify(output, null, 2));
}

main().catch(console.error);

File: src/bin/extract_diagrams.rs
//! Extract mermaid diagram examples from TypeScript test files.
//!
//! This tool parses TypeScript/JavaScript test files and extracts
//! mermaid diagram definitions from template literals.

use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Serialize, Deserialize)]
struct DiagramEntry {
    diagram: String,
    source_file: String,
    test_name: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct ExtractionResult {
    count: usize,
    diagrams: Vec<DiagramEntry>,
}

/// Diagram type keywords to look for in template literals
const DIAGRAM_TYPES: &[&str] = &[
    "sequenceDiagram",
    "flowchart",
    "graph",
    "classDiagram",
    "stateDiagram",
    "erDiagram",
    "gantt",
    "pie",
    "mindmap",
    "timeline",
    "journey",
    "quadrantChart",
    "xychart-beta",
    "sankey-beta",
    "packet-beta",
    "block-beta",
    "architecture-beta",
    "C4Context",
    "C4Container",
    "C4Component",
    "gitGraph",
    "requirementDiagram",
    "kanban",
    "info",
    "radar-beta",
    "treemap-beta",
];

fn extract_diagrams_from_file(filepath: &Path) -> Vec<DiagramEntry> {
    let content = match fs::read_to_string(filepath) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let mut diagrams = Vec::new();

    // Build pattern to match any diagram type
    let type_pattern = DIAGRAM_TYPES
        .iter()
        .map(|t| regex::escape(t))
        .collect::<Vec<_>>()
        .join("|");
    let type_regex = Regex::new(&format!("(?i){}", type_pattern)).unwrap();

    // Pattern to find template literals (backtick strings)
    let template_regex = Regex::new(r"`([^`]+)`").unwrap();

    // Pattern to find test names
    let test_regex = Regex::new(r#"(?i)(?:it|test)\s*\(\s*['"]([^'"]+)['"]"#).unwrap();

    // Track current test name by position
    let test_positions: Vec<(usize, String)> = test_regex
        .captures_iter(&content)
        .map(|cap| (cap.get(0).unwrap().start(), cap[1].to_string()))
        .collect();

    for cap in template_regex.captures_iter(&content) {
        let template_content = &cap[1];

        // Check if this template contains a diagram type
        if !type_regex.is_match(template_content) {
            continue;
        }

        let diagram_text = template_content.trim().to_string();

        // Skip if it's too short (probably a partial reference)
        if diagram_text.len() < 20 {
            continue;
        }

        // Find the test name for this diagram
        let pos = cap.get(0).unwrap().start();
        let test_name = test_positions
            .iter()
            .filter(|(p, _)| *p < pos)
            .last()
            .map(|(_, name)| name.clone())
            .unwrap_or_else(|| "unknown".to_string());

        diagrams.push(DiagramEntry {
            diagram: diagram_text,
            source_file: filepath.to_string_lossy().to_string(),
            test_name,
        });
    }

    diagrams
}

fn extract_from_directory(directory: &Path) -> Vec<DiagramEntry> {
    let mut all_diagrams = Vec::new();

    let patterns = [
        "**/*.spec.ts",
        "**/*.spec.js",
        "**/*.test.ts",
        "**/*.test.js",
    ];

    for pattern in patterns {
        let full_pattern = directory.join(pattern);
        if let Ok(paths) = glob::glob(full_pattern.to_str().unwrap_or("")) {
            for entry in paths.flatten() {
                let diagrams = extract_diagrams_from_file(&entry);
                all_diagrams.extend(diagrams);
            }
        }
    }

    all_diagrams
}

fn main() {
    let args: Vec<String> = std::env::args().collect();

    if args.len() < 2 {
        eprintln!("Usage: extract-diagrams <directory_or_file> [output.json]");
        std::process::exit(1);
    }

    let input_path = PathBuf::from(&args[1]);
    let output_path = args.get(2).map(PathBuf::from);

    let diagrams = if input_path.is_file() {
        extract_diagrams_from_file(&input_path)
    } else if input_path.is_dir() {
        extract_from_directory(&input_path)
    } else {
        eprintln!("Error: {} does not exist", input_path.display());
        std::process::exit(1);
    };

    // Remove duplicates based on diagram content
    let mut seen = HashSet::new();
    let unique_diagrams: Vec<DiagramEntry> = diagrams
        .into_iter()
        .filter(|d| {
            let key = d.diagram.trim().to_string();
            if seen.contains(&key) {
                false
            } else {
                seen.insert(key);
                true
            }
        })
        .collect();

    let result = ExtractionResult {
        count: unique_diagrams.len(),
        diagrams: unique_diagrams,
    };

    let json = serde_json::to_string_pretty(&result).unwrap();

    if let Some(output) = output_path {
        fs::write(&output, &json).expect("Failed to write output file");
        eprintln!(
            "Extracted {} diagrams to {}",
            result.count,
            output.display()
        );
    } else {
        println!("{}", json);
    }
}
